Remove CqlType in favor of ColumnType

Lorak-mmk opened the pull request:
This PR removes `CqlType` (used in schema metadata) and replaces its usages with `ColumnType` (used in statement metadata).
Initially those two structures were quite different, so most of the commits are done to make them mostly identical, which made the unification possible.

## Layout 

CqlType was split into subtypes (native type, collection type, tuple, UDT) as defined by https://cassandra.apache.org/doc/stable/cassandra/cql/types.html
ColumnType was fully inlined - it had a variant for each possible type.

I decided to change ColumnType in this regard to follow CqlType, because CqlType's approach seems to be much more elegant.

## Frozen flag

CqlType contained frozen flag, because the schema contains information about whether the collection/UDT is frozen or not.
ColumnType did not contain such flag, because this information is not present in result metadata / arguments metadata.

There wasn't any real choice - we have to store the frozen flag to accurately represent the schema metadata.
For types used in statements, this flag is simply always set to false.

## UDT shared ownership

ColumnType stored the UDT data inline, this is how the variant looked:
```rust
    UserDefinedType {
        type_name: Cow<'frame, str>,
        keyspace: Cow<'frame, str>,
        field_types: Vec<(Cow<'frame, str>, ColumnType<'frame>)>,
    },
```

CqlType stored it in a separate struct, which was held under Arc:
```rust
/// Definition of a user-defined type
    UserDefinedType {
        frozen: bool,
        // Using Arc here in order not to have many copies of the same definition
        definition: Arc<UserDefinedType>,
    },
...

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct UserDefinedType {
    pub name: String,
    pub keyspace: String,
    pub field_types: Vec<(String, CqlType)>,
}

```

This is because it may be stored in a few places:
- `Keyspace::user_defined_types`
- As a field in another UDT
- As a column in the table
- As a column in the view

This requires either cloning, or shared ownership.

## Custom type

ColumnType had a variant for Custom type, because it is defined by the protocol.
However, this type is rarely used, and we don't really have any working support for it, so I decided to remove it.

## Minor changes

I decided to change the size type of Vector to u16. It frees any users of this struct from dealing with negative numbers (which are disallowed by the protocol, and make no sense here).

Fixes https://github.com/scylladb/scylla-rust-driver/issues/691

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] ~~I added relevant tests for new features and bug fixes.~~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] ~~I have provided docstrings for the public items that I want to introduce.~~
- [ ] ~~I have adjusted the documentation in `./docs/source/`.~~ - Nothing to adjust
- [x] I added appropriate `Fixes:` annotations to PR description.

Lorak-mmk commented:
Rebased on main

Lorak-mmk commented:
Again known failure with "Failed to apply group 0 change due to concurrent modification"...

If Core is not willing to address this problem then maybe we should introduce client-side retries for those queries?

muzarski commented on scylla/src/cluster/metadata.rs:
```diff
@@ -216,7 +216,7 @@ enum PreCqlType {
         type_: Box<PreCqlType>,
         /// matches the datatype used by the java driver:
         /// <https://github.com/apache/cassandra-java-driver/blob/85bb4065098b887d2dda26eb14423ce4fc687045/core/src/main/java/com/datastax/oss/driver/api/core/type/DataTypes.java#L77>
-        dimensions: i32,
+        dimensions: u16,
```
nit: You removed the comment in `CqlType` but not in `PreCqlType`.

muzarski commented on scylla/src/cluster/metadata.rs:
```diff
@@ -275,9 +275,7 @@ pub enum CqlType {
-        /// matches the datatype used by the java driver:
-        /// <https://github.com/apache/cassandra-java-driver/blob/85bb4065098b887d2dda26eb14423ce4fc687045/core/src/main/java/com/datastax/oss/driver/api/core/type/DataTypes.java#L77>
-        dimensions: i32,
+        dimensions: u16,
```
What about `(u16::MAX, i32::MAX]` values? I know that such values are impractical and probably will never appear there, but according the specification it is possible, right?

Lorak-mmk replied:
Which specification?

- CQLv5 binary protocol only mentions vectors once:
```
5.25 vector

  For a vector of n dimensions of a fixed-length type, a sequence of those n elements.
  For a vector with variable-length elements, the size of the elements will preced
  each element.  Each element is the [bytes] representing the serialized value.  The
  number of dimensions is not encoded, since it's part of the type definition.
```
There is no defined binary type encoding for vectors (which would give is the type to use for dimensions) - it is encoded as text (custom type) which is why https://github.com/scylladb/scylla-rust-driver/pull/1165 needs to do so many changes.

- Cassandra docs say that the maximum amount of elements in a vector is 8K (2^13) which fits in u16: https://cassandra.apache.org/doc/latest/cassandra/reference/vector-data-type.html

muzarski replied:
>     * Cassandra docs say that the maximum amount of elements in a vector is 8K (2^13) which fits in u16: https://cassandra.apache.org/doc/latest/cassandra/reference/vector-data-type.html

Ok, this convinces me. I based my assumption on the code snippet (https://github.com/apache/cassandra-java-driver/blob/85bb4065098b887d2dda26eb14423ce4fc687045/core/src/main/java/com/datastax/oss/driver/api/core/type/DataTypes.java#L77). I forgot that Java does not have unsigned integral data types... But then, I believe that `short` data type in Java should be enough to store a value up to 8k. I wonder why they chose `int`.

Lorak-mmk replied:
I don't think they pay much attention to such things. If they did then CQL protocol wouldn't have such obvious issues unfixed for so long is it does now.

muzarski commented on scylla/src/client/session_test.rs:
```diff
@@ -6,7 +6,7 @@ use crate as scylla;
 use crate::batch::{Batch, BatchStatement};
```
commit `Remove CqlType and related types` - I expected this commit to contain a lot of changes. To my surprise, there are very few of them. Great job of splitting this PR into the commits - thanks to that the final replacement was trivial.

Lorak-mmk commented:
Adrressed the comments

wprzytula commented:
> Again known failure with "Failed to apply group 0 change due to concurrent modification"...
> 
> If Core is not willing to address this problem then maybe we should introduce client-side retries for those queries?

This sounds like a good workaround.

wprzytula commented on scylla/src/utils/parse.rs:
```diff
@@ -87,18 +87,18 @@ impl<'s> ParserState<'s> {
     ///
     /// An error is returned if:
-    /// * The first character is not a digit or '-'
+    /// * The first character is not a digit
     /// * The integer is larger than i32
```
:wrench: Outdated docstring: `i32` -> `u16`

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -1391,6 +1399,12 @@ pub fn deser_cql_value(
+        Vector { .. } => {
+            return Err(mk_deser_err::<CqlValue>(
+                typ,
+                BuiltinDeserializationErrorKind::Unuspported,
```
:wrench: typo: `Unuspported` -> `Unsupported`

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -1548,6 +1562,9 @@ mod test_utils {
                 Self::Collection {
                     type_: CollectionType::Set(_),
                 } => 0x0022,
+                Self::Vector { .. } => {
+                    panic!("Vector serialization unimplemented");
+                }
```
:pick: `unimplemented!()` is designed specifically for this use case.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -1582,6 +1599,9 @@ mod test_utils {
                         typ.serialize(buf)?;
                     }
                 }
+                ColumnType::Vector { .. } => {
+                    panic!("Vector serialization unimplemented");
+                }
```
:pick: ditto: `unimplemented!()`.

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -1830,6 +1830,9 @@ pub enum BuiltinDeserializationErrorKind {
     /// A deserialization failure specific to a CQL UDT.
     UdtError(UdtDeserializationErrorKind),
+
+    /// Deserialization of this CQL type is not supported by the driver
+    Unuspported,
```
:pick: Let's be consistent about dots at the end of docstrings here (and everywhere else). It matters at least for me, so perhaps for other readers too.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -52,9 +52,9 @@ pub struct TableSpec<'a> {
-    Set(Box<ColumnType<'frame>>),
+    Collection {
+        type_: CollectionType<'frame>,
+    },
```
:pick: `type_` looks awful. Let's use `typ` instead, as in many other places in the driver.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -53,13 +53,15 @@ pub struct TableSpec<'a> {
+        frozen: bool,
         definition: Arc<UserDefinedType<'frame>>,
     },
     Tuple(Vec<ColumnType<'frame>>),
```
:wrench: As this is a `pub` type, I suggest adding docstrings to all variants and their fields. Especially the `frozen` flag may require explanation, as it's only meaningful in the schema metadata case, which is not obvious.

Lorak-mmk replied:
Added documentation in a new commit.

Lorak-mmk commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -56,9 +56,7 @@ pub enum ColumnType<'frame> {
-        type_name: Cow<'frame, str>,
-        keyspace: Cow<'frame, str>,
-        field_types: Vec<(Cow<'frame, str>, ColumnType<'frame>)>,
+        definition: Arc<UserDefinedType<'frame>>,
```
That's true, but its worth noting that it's not a change from no allocations to an allocation, but from at least 1 allocation to at least 2 allocations - there was previously an allocation because of Vec.

Lorak-mmk commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -933,9 +943,11 @@ fn deser_type_generic<'frame, 'result, StrT: Into<Cow<'result, str>>>(
+                    field_types,
+                }),
             }
         }
```
Yes, that's how it has been before this PR too. This is also a pretty rare scenario I think.

Lorak-mmk commented:
Addressed all the review comments. I'll rebase on main in a separate push.

Lorak-mmk commented:
- Rebased on main
- Fixed an issue where the new types (NativeType) etc were not re-exported in `scylla` crate. Added a test for that in `hygiene.rs`

Lorak-mmk commented:
Fixed CI errors. I'll run `make  ci` again on  all commits to ensure they work.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -49,9 +49,50 @@ pub struct TableSpec<'a> {
     table_name: Cow<'a, str>,
 }
 
+/// A type of the column in schema metadata or in prepared statement.
+///
+/// In schema metadata this represents simply a column type. In prepared
+/// statement this is used in two places: as a type of bind marker and as a type
+/// of returned value.
+///
+/// Some of the variants contain a `frozen` flag. This flag is only used
+/// in schema metadata. For prepared statement bind markers and result types
+/// those fields will always be set to `false` (even if the DB column
+/// corresponding to given marker / result type is frozen).
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub enum ColumnType<'frame> {
```
:thinking: I believe the `ColumnType` is used in the result set of both prepared and unprepared statements. The docstring specifically mentions only prepared statements, which I'm afraid is misleading.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -49,9 +49,50 @@ pub struct TableSpec<'a> {
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub enum ColumnType<'frame> {
-    Custom(Cow<'frame, str>),
+    /// Types that are "simple" (non-recursive).
+    Native(NativeType),
+
+    /// Collection types: Map, Set, and List. Those are composite types with
+    /// dynamic size but constant predefined element types.
+    Collection {
+        frozen: bool,
+        typ: CollectionType<'frame>,
+    },
+
+    // A composite list-like type that has a defined size and all its elements
+    // are of the same type. Intuitively, it can be viewed as a list with constant
+    // predefined size, or as a tuple which has all elements of the same type.
+    Vector {
+        typ: Box<ColumnType<'frame>>,
+        dimensions: u16,
+    },
+
+    // A C-struct-like type defined by the user.
+    UserDefinedType {
+        frozen: bool,
+        definition: Arc<UserDefinedType<'frame>>,
+    },
+
+    // A composite type with a defined size and elements of possibly different,
+    // but predefined, types.
+    Tuple(Vec<ColumnType<'frame>>),
```
:+1: Very useful docstrings, good job!

:wrench: The lower three variants have `//` instead of `///`.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -66,68 +107,85 @@ pub enum ColumnType<'frame> {
     Varint,
 }
 
+/// Collection variants of [ColumnType]. A collection is a composite type that:
+/// has no predefined size, so it is possible to add and remove values to/from it.
+///
+/// Tuple and vector are not collections because they don't have dynamic size.
```
:pick: I'd remove negations from both sentences:

```suggestion
/// Collection variants of [ColumnType]. A collection is a composite type that:
/// has dynamic size, so it is possible to add and remove values to/from it.
///
/// Tuple and vector are not collections because they have predefined size.
```

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -1269,109 +1342,116 @@ pub fn deser_cql_value(
     let v = Some(FrameSlice::new_borrowed(buf));
```
:pushpin: When removing the legacy deserialization framework, please check if it's possible to remove the workaround in `FrameSlice`: the `new_borrowed` constructor.
